### PR TITLE
Added compiler

### PR DESCRIPTION
**Pull request by DimitarSD**

### CODE REVIEW

File: cli-commands/commands.js [hunk 1]
```diff
@@ -61,6 +62,20 @@ const commands = [
 		commandProcessor: (argv) => {
 			history.run(argv.limit, console);
 		}
+	},
+	{
+		command: 'compile [default_path]',
```
Comment (Perseverance) [inline cli-commands/commands.js]: `default_path` is not the correct name contextually. It should be just `dir`.

File: cli-commands/commands.js [hunk 2]
```diff
@@ -61,6 +62,20 @@ const commands = [
...
+		description: 'Compiles smart contracts',
```
Comment (Perseverance) [inline cli-commands/commands.js]: I think this should be better described. Like:
Compiles the smart contracts that are in the directory `contracts` in the path provided by the `dir` parameter (defaults to `.`).

File: cli-commands/commands.js [hunk 3]
```diff
@@ -61,6 +62,20 @@ const commands = [
...
+		argumentsProcessor: (yargs) => {
+			yargs.positional('default_path', {
+				describe: 'Specifies the default path to all folders',
```
Comment (Perseverance) [inline cli-commands/commands.js]: Correct this message too

File: cli-commands/compiler/compiler.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
+const logsStore = require('./../../logs-store/logs-store');
```
Comment (Perseverance) [inline cli-commands/compiler/compiler.js]: You do not need the logs store

File: cli-commands/compiler/compiler.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+const utils = require('./../util');
+
+let compiler = require("./etherlime-workflow-compile/index");
+let Resolver = require("./etherlime-resolver/index");
+
+const run = async (defaultPath) => {
+    const initialRecordsCount = logsStore.getHistory().length;
```
Comment (Perseverance) [inline cli-commands/compiler/compiler.js]: You do not need to print report

File: cli-commands/compiler/compiler.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        console.log(`Your compile script finished successfully!`);
+    } catch (e) {
+        if (!silent) {
+            console.error(e);
+        }
+
+        console.log(`Your compile script finished with failure!`);
+    }
+
+    const records = logsStore.getHistory();
```
Comment (Perseverance) [inline cli-commands/compiler/compiler.js]: You do not need history. Too much copy pasting :D

File: cli-commands/compiler/compiler.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
+const logsStore = require('./../../logs-store/logs-store');
+const utils = require('./../util');
+
+let compiler = require("./etherlime-workflow-compile/index");
+let Resolver = require("./etherlime-resolver/index");
+
+const run = async (defaultPath) => {
+    const initialRecordsCount = logsStore.getHistory().length;
+
+    try {
```
Comment (Perseverance) [inline cli-commands/compiler/compiler.js]: Remove the try catch

File: cli-commands/compiler/compiler.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+        let resolverOptions = {
```
Comment (Perseverance) [inline cli-commands/compiler/compiler.js]: Move the compilation with options into another function called `performCompilation`

File: cli-commands/compiler/compiler.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        };
+
+        Resolver(resolverOptions);
+
+        let compileOptions = {
+            "contracts_directory": `${defaultPath}/contracts`,
+            "contracts_build_directory": `${defaultPath}/build`
+        };
+
+        compiler.compile(compileOptions, () => {});
```
Comment (Perseverance) [inline cli-commands/compiler/compiler.js]: What is the second param? Why empty function?

File: cli-commands/init/init.js [hunk 10]
```diff
@@ -22,6 +23,12 @@ const copyDeployFile = (libraryDirectory) => {
 	fs.copyFileSync(deploymentFileSource, deploymentFileDestination);
 }
 
+const createContractsFolder = () => {
```
Comment (Perseverance) [inline cli-commands/init/init.js]: Lets add console log for creating contracts file structure

File: package.json [hunk 11]
```diff
@@ -19,10 +19,22 @@
...
+    "async-each-series": "^1.1.0",
+    "async.whilst": "^0.5.2",
+    "colors": "^1.3.0",
+    "ethjs-abi": "^0.2.1",
+    "graphlib": "^2.1.5",
+    "mkdirp": "^0.5.1",
+    "node-dir": "^0.1.17",
+    "original-require": "^1.0.1",
+    "solc": "^0.4.24",
+    "web3": "0.20.6"
```
Comment (Perseverance) [inline package.json]: We should strive to remove web3 and ethjs

File: cli-commands/commands.js [hunk 12]
```diff
@@ -61,6 +62,20 @@ const commands = [
 		commandProcessor: (argv) => {
 			history.run(argv.limit, console);
 		}
+	},
+	{
+		command: 'compile [dir]',
+		description: 'Compiles the smart contracts that are in the directory contracts in the path provided by the dir parameter (defaults to .)',
+		argumentsProcessor: (yargs) => {
+			yargs.positional('dir', {
+				describe: 'Specifies the default dir to all folders',
```
Comment (Perseverance) [inline cli-commands/commands.js]: Lets write this as `specifies the root dir to read the contracts and place the build folder`

File: cli-commands/compiler/etherlime-artifactor/index.js [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    var output_path = object.contractName;
+
+    // Create new path off of destination.
+    output_path = path.join(self.destination, output_path);
+    output_path = path.resolve(output_path);
+
+    // Add json extension.
+    output_path = output_path + ".json";
```
Comment (Perseverance) [inline cli-commands/compiler/etherlime-artifactor/index.js]: You can do this with string literal

File: cli-commands/compiler/compiler.js [hunk 14]
```diff
@@ -0,0 +1,28 @@
+const utils = require('./../util');
+
+let compiler = require("./etherlime-workflow-compile/index");
+let Resolver = require("./etherlime-resolver/index");
+
+const performCompilation = () => {
+    let resolverOptions = {
+        "working_directory": `${defaultPath}/contracts`,
```
Comment (Perseverance) [inline cli-commands/compiler/compiler.js]: Where does defaultPath comes from? It should be a param but I do not see it passed

File: cli-commands/compiler/etherlime-blockchain-utils/index.js [hunk 15]
```diff
@@ -0,0 +1,65 @@
+var Blockchain = {
```
Comment (Perseverance) [inline cli-commands/compiler/etherlime-blockchain-utils/index.js]: Do we need this file and why?

File: package.json [hunk 16]
```diff
@@ -19,10 +19,22 @@
...
     "moment": "^2.22.2",
     "readline-sync": "^1.4.9",
-    "yargs": "^11.0.0"
+    "yargs": "^11.0.0",
+    "ajv": "^6.5.2",
+    "async": "^2.6.1",
+    "async-each-series": "^1.1.0",
+    "async.whilst": "^0.5.2",
+    "colors": "^1.3.0",
+    "ethjs-abi": "^0.2.1",
```
Comment (Perseverance) [inline package.json]: Can we remove ethjs-abi

File: cli-commands/compiler/etherlime-blockchain-utils/index.js (hunk 15 above)
Comment (DimitarSD) [reply cli-commands/compiler/etherlime-blockchain-utils/index.js]: Yes, we need it. It's used in contract.js.

File: package.json (hunk 16 above)
Comment (DimitarSD) [reply package.json]: I think we can't. It's used in contract.js and it's a part of the whole workflow

File: cli-commands/compiler/compiler.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
+let compiler = require("./etherlime-workflow-compile/index");
+let Resolver = require("./etherlime-resolver/index");
+
+const performCompilation = (defaultPath) => {
+    let resolverOptions = {
+        "working_directory": `${defaultPath}/contracts`,
```
Comment (Perseverance) [inline cli-commands/compiler/compiler.js]: You probably do not need the " around the keys.

**Pull request merged**